src/app-layer-htp.c

```diff
@@ -975,29 +976,52 @@ static AppLayerResult HTPHandleResponseData(Flow *f, void *htp_state, AppLayerPa
                 if (tx != NULL && tx->response_status_number == 101) {
                     htp_header_t *h =
                             (htp_header_t *)htp_table_get_c(tx->response_headers, "Upgrade");
-                    if (h == NULL || bstr_cmp_c(h->value, "h2c") != 0) {
-                        break;
-                    }
-                    if (AppLayerProtoDetectGetProtoName(ALPROTO_HTTP2) == NULL) {
-                        // if HTTP2 is disabled, keep the HTP_STREAM_TUNNEL mode
+                    if (h == NULL) {
                         break;
                     }
                     uint16_t dp = 0;
                     if (tx->request_port_number != -1) {
                         dp = (uint16_t)tx->request_port_number;
                     }
                     consumed = htp_connp_res_data_consumed(hstate->connp);
-                    hstate->slice = NULL;
-                    if (!AppLayerRequestProtocolChange(hstate->f, dp, ALPROTO_HTTP2)) {
-                        HTPSetEvent(hstate, NULL, STREAM_TOCLIENT,
-                                HTTP_DECODER_EVENT_FAILED_PROTOCOL_CHANGE);
-                    }
-                    // During HTTP2 upgrade, we may consume the HTTP1 part of the data
-                    // and we need to parser the remaining part with HTTP2
-                    if (consumed > 0 && consumed < input_len) {
-                        SCReturnStruct(APP_LAYER_INCOMPLETE(consumed, input_len - consumed));
+                    if (bstr_cmp_c(h->value, "h2c") == 0) {
+                        if (AppLayerProtoDetectGetProtoName(ALPROTO_HTTP2) == NULL) {
+                            // if HTTP2 is disabled, keep the HTP_STREAM_TUNNEL mode
+                            break;
+                        }
+                        hstate->slice = NULL;
+                        if (!AppLayerRequestProtocolChange(hstate->f, dp, ALPROTO_HTTP2)) {
+                            HTPSetEvent(hstate, NULL, STREAM_TOCLIENT,
+                                    HTTP_DECODER_EVENT_FAILED_PROTOCOL_CHANGE);
+                        }
+                        // During HTTP2 upgrade, we may consume the HTTP1 part of the data
+                        // and we need to parser the remaining part with HTTP2
+                        if (consumed > 0 && consumed < input_len) {
+                            SCReturnStruct(APP_LAYER_INCOMPLETE(consumed, input_len - consumed));
+                        }
+                        SCReturnStruct(APP_LAYER_OK);
+                    } else if (bstr_cmp_c_nocase(h->value, "WebSocket") == 0) {
+                        if (AppLayerProtoDetectGetProtoName(ALPROTO_WEBSOCKETS) == NULL) {
+                            // if WS is disabled, keep the HTP_STREAM_TUNNEL mode
+                            break;
+                        }
+                        hstate->slice = NULL;
+                        if (AppLayerExpectationCreate(f, STREAM_TOCLIENT | STREAM_TOSERVER, f->sp,
```

Comment:
surprised to see this. Expectations are about related but unique flows (so different 5 tuple), that is not the case here, right?

Reply:
Expectations create a new flow, so shouldn't be used here. It's the wrong API for the problem.

Forcing the protocol change, perhaps with some validation that it indeed looks like websock, sounds better.

Reply:
Now, that looks evil 👿 
Thanks for the explanation :-)

Reply:
Got a new hack for next PR iteration